### ISSUE TITLE
A way to shorten/alias method name derived queries. [DATACMNS-1213]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-commons#1652]
**[justin mccune](https://jira.spring.io/secure/ViewProfile.jspa?name=justin.mccune@gmail.com)** opened **[DATACMNS-1213](https://jira.spring.io/browse/DATACMNS-1213?redirect=false)** and commented

Many think that it's great to use the name of the method to infer the SQL for a query on a repository.  However, often the use of descriptive field names and one or two concerns (an AND operation with a SORT ordering) leads to incredibly long (or "ugly" names). For example one properly named long field and a simple boolean & sort order lead to the repository method name being 96 characters long. This destroys many style guides line-lengths when used.

While it is an alternative to use `@Query` make a shorter method name  and write the actual query by hand, or to use java's "default" on interface to alias a shorter form that uses the full / ugly query.  They both are not as elegant as the simplicity of the named query where preferably I alias it.

Often the alias / shorter name is just as meaningful and clear in meaning even if omitting implicit/assumed context-- see examples below.

This has come up several times with coworkers (in multiple companies) & a quick search has shown that others have the same concern.

https://stackoverflow.com/questions/30019408/how-to-shorten-names-of-query-methods-in-spring-data-jpa-repositories

When I read the above, I thought perhaps I'd found the solution-- but it was a suggestion that I couldn't find in the Spring Data feature request list, so I thought I'd add it.

So what I'd like would be something like the below-- where the annotation name is used instead of the actual method name to build the query.

```
@QueryByMethodName("findOneByDeletedIsFalseAndEmail")
User findOneByEmail(String email);

@QueryByMethodName("findOneByReallyLongFieldNameAndActiveStatusIsTrue")
Entity findOneByRLFN(ReallyLongFieldNameType value)
```

This improves both simplicity and readability.  And I'm not particular about the annotation name.

---

16 votes, 8 watchers

**Comment by gbraleigh** [comment]
Yes, please! This seems like a relatively simple feature that would be very valuable, improving our ability to read and easily understand repository generated method names.

**Comment by kadirtn** [comment]
Method name gets too crowd just for a simple query, proposed approach seems convenient.

**Comment by grzechup** [comment]
I would love to see this feature. For example, I have the following method query in my repository.
`    Optional<CallManagerMonitoredUser> findByServiceProviderBwIdAndGroupBwIdAndUserBwIdAndMonitoredContactContactNameAndMonitoredContactContactNumber(String serviceProviderId,
                                                                                                                                                      String groupId,
                                                                                                                                                      String userBwId,
                                                                                                                                                      String contactName,
                                                                                                                                                      String contactNumber);`
That will add much readability to my code.

**Comment by bektun** [comment]
push

**Comment by bektun** [comment]
if you extend your repository with `JpaRepository` then you can use a default method which calls your origin repository method:
`public interface UserRepository extens JpaRepositry<String, User>
   List<User> findUserByCreatedDateAndHairColorAndFavoriteTeamAndBLaBla(param1, param2...);
   default findBlaBlaUser(param1, param2...) {
      return findUserByCreatedDateAndHairColorAndFavoriteTeamAndBLaBla(param1, param2...);
`
in your service you can now call myUserRepo.findBlaBlaUser(param1, param2...)

**Comment by mp911de** [comment]
Spring Data is not in charge of defining yet another query language. We've built query derivation to allow users conveniently declare simple queries without having to dive into the store-specific query language to implement common use cases.
If a method becomes too long, then it no longer fits the intended usage pattern, and it falls out of this category. 

Putting the potential `String` into an annotation moves the problem from the method declaration to an annotation. Still, it doesn't cure the underlying problem of fixing a query that is too complex. If it is complex, then it needs to be expressed that way or be addressed using the store's query language; We recommend either using `@Query` or a custom implementation. 

Furthermore, we are strictly against putting code into annotations that actually belong somewhere else. Such a change leads to removing information from the call site. The method name becomes less expressive at its best. In the worst case, it contradicts what was specified in an annotation.

**Comment by jxblum** [comment]
I speak here in the general case, and not specifically with respect to SD JPA.

Perhaps a less-known feature of Spring Data, and unfortunately, it is not a feature that is commonly supported across all Spring Data modules, but a nice alternative none-the-less is, some Spring Data modules support a [`QueryLookupStrategy`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/query/QueryLookupStrategy.html) that pulls the actual store-specific query from a Java Properties file and uses the entity's simple type name + the query method's name as a key to lookup the query to use for that _Repository_ method when invoked.

> TIP: Query Lookup Strategies are loosely described in the Spring Data Commons reference documentation, [here](https://docs.spring.io/spring-data/commons/docs/current/reference/html/#repositories.query-methods.details). The key statements in this documentation are 1) "_Some strategies may not be supported for particular datastores._" and 2) under `USE_DECLARED_QUERY` option, "_The query can be defined by an annotation somewhere **or declared by other means**. See the documentation of the specific store to find available options for that store._"

Building on my statement above with a concrete example, suppose you have a _Repository_ with the following query method:

```java
interface CustomerRepository extends CrudRepository<Customer, Long> {

  Optional<Customer> findVipCustomers(String firstName, String lastName);

}
```

Then, the actual store-specific query could be defined in a Java Properties file (pseudo syntax used):

```properties
// Properties defining store-specific queries for 
Customer.findByName=SELECT DISTINCT c.first_name, c.last_name, c.birth_date \
   FROM Customers c \
   WHERE c.first_name = ? \
   AND c.last_name = ? \
  AND c.vip = true \
  ORDER BY c.last_name ASC, c.birth_date DESC
```

Typically, the Java Properties file location/name is in `META-INF/<data-store-name>-named-queries.properties`, where `<data-store-name>` could be "cassandra". Of course, this location/name can be changed, but is again store specific.

To see how this works in _Spring Data Repository_ infrastructure, there are few things you have to know.

Internally, Spring Data (Commons) represents these externally sourced (e.g. Java Properties) and named queries with the [`NamedQueries`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/core/NamedQueries.html) type. 1 such implementation is the [`PropertiesBasedNamedQueries`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/core/support/PropertiesBasedNamedQueries.html) type.

The "name" of the query is derived from the [`QueryMethod`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/query/QueryMethod.html) type representing the query method (that represents the derived query) declared in your application _Repository_ interface (e.g. `CustomerRepository`, such as `findVipCustomers(..)`). Specifically, Spring Data's _Repository_ infrastructure calls the [`QueryMethod.getNamedQueryName()`](https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/query/QueryMethod.html#getNamedQueryName()) method to resolve the "name" given to the query (or key in the Java Properties file). The [default implementation](https://github.com/spring-projects/spring-data-commons/blob/3.0.2/src/main/java/org/springframework/data/repository/query/QueryMethod.java#L153-L160) combines the **entity simple type name** with the actual **Java `Method` name** of the derived query method declare in the _Repository_. AFAIK, most SD modules do not override the default implementation.

The "location" of this Java Properties file defining the queries can be changed with _Repository_ enabling annotation for the specific data store in use. For example, when using Spring Data Cassandra, you would use the `@EnableCassandraRepositories` annotation, [`namedQueriesLocation`](https://docs.spring.io/spring-data/cassandra/docs/current/api/org/springframework/data/cassandra/repository/config/EnableCassandraRepositories.html#namedQueriesLocation()) attribute.

Then, internally, and by example, Spring Data Cassandra's [`QueryLookupStrategy`](https://github.com/spring-projects/spring-data-cassandra/blob/4.0.2/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/repository/support/CassandraRepositoryFactory.java#L109-L145) will use the above mentioned infrastructure objects to [lookup the query from an external source](https://github.com/spring-projects/spring-data-cassandra/blob/4.0.2/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/repository/support/CassandraRepositoryFactory.java#L133-L136), such as a Java Properties file (default).

It is pretty eloquent.

It seems to me, Spring Data JPA's [`QueryLookupStrategy`](https://github.com/spring-projects/spring-data-jpa/blob/3.0.2/spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/JpaQueryLookupStrategy.java) supports named query resolution quite nicely.  See [here](https://github.com/spring-projects/spring-data-jpa/blob/3.0.2/spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/JpaQueryLookupStrategy.java#L128-L185) in particular.

With all the mucking with [derived] query method names (and overly used `@Query` annotations for rather complex queries), I am not sure why users don't use this Spring Data feature more.

**Comment by jxblum** [comment]
To summarize the different approaches for declaring queries with Spring Data's Repository infrastructure, approaches include the following common strategies:

1) Queries derived from a [carefully crafted and named _Repository_ query methods](https://docs.spring.io/spring-data/commons/docs/current/reference/html/#repositories.query-methods.query-creation).

2) Defining queries with the [`@Query`](https://docs.spring.io/spring-data/jpa/docs/current/api/org/springframework/data/jpa/repository/Query.html) annotation declared on _Repository_ query methods.

3) [Custom Repository method implementations](https://docs.spring.io/spring-data/commons/docs/current/reference/html/#repositories.custom-implementations).

4) Externally managed and defined queries as I [described above](https://github.com/spring-projects/spring-data-commons/issues/1652#issuecomment-1442474080).

Of course, some Spring Data modules may support other forms for querying as well, including, but not limited to: [_Query by Example_](https://docs.spring.io/spring-data/commons/docs/current/reference/html/#query-by-example) or [_Querydsl_](http://querydsl.com/), JOOQ, HQL, etc.

**Comment by Mooninaut** [comment]
> Spring Data is not in charge of defining yet another query language. We've built query derivation to allow users conveniently declare simple queries without having to dive into the store-specific query language to implement common use cases. If a method becomes too long, then it no longer fits the intended usage pattern, and it falls out of this category.

Except query derivation *is* a query DSL. I can't see how to reconcile these two statements.

> Putting the potential `String` into an annotation moves the problem from the method declaration to an annotation. Still, it doesn't cure the underlying problem of fixing a query that is too complex. If it is complex, then it needs to be expressed that way or be addressed using the store's query language; We recommend either using `@Query` or a custom implementation.

Instead of moving the problem to an annotation (`@QueryByMethodName`), move the problem to a more complicated and harder to read annotation (`@Query`)?

> Furthermore, we are strictly against putting code into annotations that actually belong somewhere else. Such a change leads to removing information from the call site. The method name becomes less expressive at its best. In the worst case, it contradicts what was specified in an annotation.

TLDR: I don't understand rejecting `@QueryByMethodName` with a rationale that applies even more strongly to `@Query`.

I write this because I have observed **in practice**, that moving logic from query derivation to `@Query` dramatically increases the chance that the resulting query logic will accidentally differ from the intended logic. The derivation DSL, like it or not, is a valuable tool to express simple queries simply. The issue is not that the resulting queries are too *complex* but that their method names are *too long* when the queries have multiple parameters.

**Comment by odrotbohm** [comment]
> Except query derivation is a query DSL. I can't see how to reconcile these two statements.

It might be a language thing, but “yet another” intended to mean “another one on top of the already existing (query derivation, store specific queries, Querydsl, query by example, custom implementations) ones”.

Generally speaking, for each request, we have to judge different aspects:

1.  Benefit of the change to the user
2. Effects on the complexity to the maintainers (both implementation and documentation)
3. Effects on the cognitive complexity for the users (to understand the tradeoff, when to use what etc.)

Given that all these alternatives already exist, 1 is low by definition. 2 and 3 definitely increase, and we came to the conclusion that adding what's asked for is not worth the effort. Rest assured that we have a pretty good impression of how our users use repository query methods.
